File: quizzes5/15-problem.ts
export {};

const rotate = (args: number[], k: number): number[] => {
  const n = args.length;
  if (n <= 1) return args;

  k = k % n;
  if (k === 0) return args;

  const reverse = (arr: number[], start: number, end: number) => {
    while (start < end) {
      [arr[start], arr[end]] = [arr[end], arr[start]];
      start++;
      end--;
    }
  };

  reverse(args, 0, n - 1);
  reverse(args, 0, k - 1);
  reverse(args, k, n - 1);

  return args;
};

const array: number[] = [1, 2, 3, 4, 5, 6, 7];
console.log(JSON.stringify({ data: rotate(array, 3) }, null, 2));
// Output: [5, 6, 7, 1, 2, 3, 4]

File: bins/binsV2/bin/bins2/15-problem.rb
# frozen_string_literal: true

require 'json'

def rotate(args, k)
  n = args.length
  return args if n <= 2

  k = k % n
  return args if k === 0

  def reverse(arr, start, stop)
    while start < stop
      arr[start], arr[stop] = arr[stop], arr[start]
      start += 1
      stop -= 1
    end
  end

  reverse(args, 0, n - 1)
  reverse(args, 0, k - 1)
  reverse(args, k, n - 1)

  args
end

array = [ 1, 2, 3, 4, 5, 6, 7 ]
puts JSON.pretty_generate({ data: rotate(array, 3) })
# Output: [5, 6, 7, 1, 2, 3, 4]
